lightning/src/offers/static_invoice.rs

```diff
@@ -0,0 +1,359 @@
+	(162, blindedpay: (Vec<BlindedPayInfo>, WithoutLength, Iterable<'a, BlindedPayInfoIter<'a>, BlindedPayInfo>)),
+	(164, created_at: (u64, HighZeroBytesDroppedBigSize)),
+	(166, relative_expiry: (u32, HighZeroBytesDroppedBigSize)),
+	(172, fallbacks: (Vec<FallbackAddress>, WithoutLength)),
+	(174, features: (Bolt12InvoiceFeatures, WithoutLength)),
+	(176, node_id: PublicKey),
+	(178, message_paths: (Vec<BlindedPath>, WithoutLength)),
+});
+
+type FullInvoiceTlvStream = (OfferTlvStream, InvoiceTlvStream, SignatureTlvStream);
+
+impl SeekReadable for FullInvoiceTlvStream {
+	fn read<R: io::Read + io::Seek>(r: &mut R) -> Result<Self, DecodeError> {
+		let offer = SeekReadable::read(r)?;
+		let invoice = SeekReadable::read(r)?;
+		let signature = SeekReadable::read(r)?;
+
+		Ok((offer, invoice, signature))
+	}
+}
+
+impl TryFrom<ParsedMessage<FullInvoiceTlvStream>> for StaticInvoice {
+	type Error = Bolt12ParseError;
+
+	fn try_from(invoice: ParsedMessage<FullInvoiceTlvStream>) -> Result<Self, Self::Error> {
+		let ParsedMessage { bytes, tlv_stream } = invoice;
+		let (offer_tlv_stream, invoice_tlv_stream, SignatureTlvStream { signature }) = tlv_stream;
+		let contents = InvoiceContents::try_from((offer_tlv_stream, invoice_tlv_stream))?;
+
+		let signature = match signature {
+			None => {
+				return Err(Bolt12ParseError::InvalidSemantics(
+					Bolt12SemanticError::MissingSignature,
+				))
+			},
+			Some(signature) => signature,
+		};
+		let tagged_hash = TaggedHash::from_valid_tlv_stream_bytes(SIGNATURE_TAG, &bytes);
+		let pubkey = contents.signing_pubkey;
+		merkle::verify_signature(&signature, &tagged_hash, pubkey)?;
+
+		Ok(StaticInvoice { bytes, contents, signature })
+	}
+}
+
+impl TryFrom<(OfferTlvStream, InvoiceTlvStream)> for InvoiceContents {
+	type Error = Bolt12SemanticError;
+
+	fn try_from(tlv_stream: (OfferTlvStream, InvoiceTlvStream)) -> Result<Self, Self::Error> {
+		let (
+			offer_tlv_stream,
+			InvoiceTlvStream {
+				payment_paths,
+				blindedpay,
+				created_at,
+				relative_expiry,
+				fallbacks,
+				features,
+				node_id,
+				message_paths,
+			},
+		) = tlv_stream;
+
+		let payment_paths = construct_payment_paths(blindedpay, payment_paths)?;
+		let message_paths = message_paths.ok_or(Bolt12SemanticError::MissingPaths)?;
+
+		let created_at = match created_at {
+			None => return Err(Bolt12SemanticError::MissingCreationTime),
+			Some(timestamp) => Duration::from_secs(timestamp),
+		};
+
+		let relative_expiry = relative_expiry.map(Into::<u64>::into).map(Duration::from_secs);
+
+		let features = features.unwrap_or_else(Bolt12InvoiceFeatures::empty);
+
+		let signing_pubkey = match node_id {
+			None => return Err(Bolt12SemanticError::MissingSigningPubkey),
+			Some(node_id) => {
+				let offer_node_id =
+					offer_tlv_stream.node_id.ok_or(Bolt12SemanticError::MissingSigningPubkey)?;
+				if node_id != offer_node_id {
+					return Err(Bolt12SemanticError::InvalidSigningPubkey);
+				}
+				node_id
+			},
+		};
```

Comment:
Looks like this will need to account for the recent spec change that allows leaving out `offer_tlv_stream.node_id` if a blinded path is included.